pkg/morph/deploy/nns.go

```diff
@@ -226,3 +227,174 @@ func lookupNNSDomainRecord(inv *invoker.Invoker, nnsContract util.Uint160, domai
+
+	onNotaryDepositDeficiency notaryDepositDeficiencyHandler
+}
+
+// updateNNSContract synchronizes on-chain NNS contract (its presence is a
+// precondition) with the local one represented by compiled executables. If
+// on-chain version is greater or equal to the local one, nothing happens.
+// Otherwise, transaction calling 'update' method is sent.
+//
+// Local manifest is extended with committee group represented by the
+// parameterized private key.
+//
+// Function behaves similar to initNNSContract in terms of context.
+func updateNNSContract(ctx context.Context, prm updateNNSContractPrm) error {
+	bLocalNEF, err := prm.localNEF.Bytes()
+	if err != nil {
+		// not really expected
+		return fmt.Errorf("encode local NEF of the NNS contract into binary: %w", err)
+	}
+
+	jLocalManifest, err := json.Marshal(prm.localManifest)
+	if err != nil {
+		// not really expected
+		return fmt.Errorf("encode local manifest of the NNS contract into JSON: %w", err)
+	}
+
+	committeeActor, err := newCommitteeNotaryActor(prm.blockchain, prm.localAcc, prm.committee)
+	if err != nil {
+		return fmt.Errorf("create Notary service client sending transactions to be signed by the committee: %w", err)
+	}
+
+	var updateTxValidUntilBlock uint32
+
+	for ; ; prm.monitor.waitForNextBlock(ctx) {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("wait for NNS contract synchronization: %w", ctx.Err())
+		default:
+		}
+
+		prm.logger.Info("reading on-chain state of the NNS contract...")
+
+		nnsOnChainState, err := readNNSOnChainState(prm.blockchain)
+		if err != nil {
+			prm.logger.Error("failed to read on-chain state of the NNS contract, will try again later", zap.Error(err))
+			continue
+		} else if nnsOnChainState == nil {
+			// NNS contract must be deployed at this stage
+			return errors.New("missing required NNS contract on the chain")
+		}
+
+		if nnsOnChainState.NEF.Checksum == prm.localNEF.Checksum {
+			// manifests may differ, but currently we should bump internal contract version
+			// (i.e. change NEF) to make such updates. Right now they are not supported due
+			// to dubious practical need
+			// Track https://github.com/nspcc-dev/neofs-contract/issues/340
+			prm.logger.Info("same local and on-chain checksums of the NNS contract NEF, update is not needed")
+			return nil
+		}
+
+		prm.logger.Info("NEF checksums of the on-chain and local NNS contracts differ, need an update")
+
+		versionOnChain, err := readContractOnChainVersion(prm.blockchain, nnsOnChainState.Hash)
+		if err != nil {
+			prm.logger.Error("failed to read on-chain version of the NNS contract, will try again later", zap.Error(err))
+			continue
+		}
+
+		// we could also try to compare on-chain version with the local one and tune
+		// update strategy: don't try to update with earlier version and catch updated
+		// contracts with unchanged version (blunder, we should react to it). For
+		// simplicity, we naively rely on the version change when the contract changes
```

Comment:
we can't catch this because one contract on the chain, and the other is local - I can build application with any contract. So, in practice we will TRY to upgrade versions properly, but in theory everything is possible

Reply:
we can encounter case when NEF changed but version not bumped. Strictly speaking, it's a failure. And it's worth to catch this case. That's what the comment about

Reply:
> we can encounter case when NEF changed but version not bumped

Right now we can't, and that's what I'm talking about. Doany of our deployment/update procedurs assume that this can happen? I doubt so.